torchao/prototype/awq/core.py

```diff
@@ -0,0 +1,149 @@
+from dataclasses import dataclass
+from typing import Tuple, Optional
+
+import torch
+import torch.nn.functional as F
+
+from torchao.dtypes.uintx.Uintx import to_uintx
+from torchao.dtypes.affine_quantized_tensor import (
+    to_affine_quantized_intx,
+    LayoutType,
+    register_layout_cls,
+    PlainAQTLayout,
+    register_aqt_quantized_linear_dispatch
+
+) 
+from torchao.quantization.quant_primitives import (
+    MappingType,
+    ZeroPointDomain,
+)
+from torchao.quantization.observer import (
+    AffineQuantizedObserverBase,
+)
+
+
+class AWQObserver(AffineQuantizedObserverBase):
+    def __init__(self,
+        weight: torch.Tensor,
+        bias: torch.Tensor,
+        block_size: Tuple,
+        input_dtype: torch.dtype,
+        mapping_type: MappingType,
+        target_dtype: torch.dtype,
+        device: str,
+        scale_search_space_size: int = 20,
+        quant_min: Optional[int] = None,
+        quant_max: Optional[int] = None,
+        eps: Optional[float] = None,
+        scale_dtype: Optional[torch.dtype] = None,
+        zero_point_dtype: Optional[torch.dtype] = None,
+        preserve_zero: Optional[bool] = True,
+        zero_point_domain = ZeroPointDomain.INT,
+    ):
+        super().__init__(
+            mapping_type,
+            target_dtype,
+            block_size = block_size, 
+            quant_min = quant_min,
+            quant_max = quant_max,
+            eps = eps,
+            scale_dtype = scale_dtype,
+            zero_point_dtype = zero_point_dtype,
+            preserve_zero = preserve_zero,
+            zero_point_domain = zero_point_domain,
+        )
+        self.weight = weight
+        self.bias = bias
+        self.scale_options = scale_search_space_size
+        self.scales = None
+        self.device = device
+
+    @torch.no_grad()
+    def forward(self, input: torch.Tensor, output: torch.Tensor):
+        average = input.abs().view(-1,input.shape[-1]).mean(0)
+        
+        best_loss = float('inf')
+        scaleopts = []
+        for i in range(self.scale_options):
+            ratio = i * 1 / self.scale_options
+            scales = average.pow(ratio)
+            scales = scales / (scales.max() * scales.min()).sqrt()
+            layout = AWQLayoutType(scales, self.target_dtype)
+            tensor_dtype = torch.int8 if self.target_dtype == torch.int8 else torch.uint8
```

Comment:
are you using torch.uint8 + quant_min/quant_max or just using native torch.uint1-torch.uint7 dtypes? seems both are used

Reply:
Switched it to using torch.uint1-8 and setting quant_max accordingly